Adds section for Ken Burns Effect

Enriqe opened the pull request:
Adds a section to amp-story describing how to achieve a "Ken Burns" effect using new animation presets.

Enriqe commented:
/cc @newmuis

sebastianbenz commented on src/stories/20_Features/Animations.html:
```diff
@@ -212,6 +215,50 @@ <h1>whoosh-in-right</h1>
       </amp-story-grid-layer>
     </amp-story-page>
+
+    <!-- ## Ken Burns Effect -->
```
You don't need to put every line into a separate comment, something like this should work:

```
<!--- 
## Ken Burns

You can achieve a "Ken Burns" effect by combining the 'pan' and 'zoom' animations. Here's how to do it:

* ...
* ...

-->

sebastianbenz commented on src/stories/20_Features/Animations.html:
```diff
@@ -212,6 +215,50 @@ <h1>whoosh-in-right</h1>
+              height="768"
+              animate-in="pan-down"
+              animate-in-duration="30s"
+              animate-in-after="ken-burns-img1"
```
`</amp-img>` missing

Enriqe commented:
> Very cool! Is it supposed to work/validate yet?

As per #14157 it is now going to be validated :)

Moved the effect to a new section like you suggested and added separate samples for the pan and zoom animations. Let me know what you think!

sebastianbenz commented on src/stories/20_Features/Visual_Effects.html:
```diff
@@ -0,0 +1,106 @@
+<!--
```
Sorry for not being clear enough. I meant a new category "Visual Effects" and the sample would be named Ken Burns. You can do this by renaming the file to:

```
src/stories/30_Visual_Effects/Ken_Burns.html
```

Creating a new directory (30_Visual_Effects) will create a new category. You can copy and adjust the index.json file from one of the other directories (it contains the category description).

sebastianbenz commented on src/stories/20_Features/Visual_Effects.html:
```diff
@@ -0,0 +1,106 @@
+  <amp-story standalone>
+    <!-- 
+    ## Ken Burns Effect
+    Best known for his work on documentaries, Ken Burns developed a signature effect in his work which consisted of a combination of panning and zooming over an image. 
```
This paragraph would then become the new introduction.

sebastianbenz commented on src/stories/20_Features/Visual_Effects.html:
```diff
@@ -0,0 +1,106 @@
+    We can achieve this simple yet popular and captivating effect by combining the 'pan' and 'zoom' animations in `amp-story`. Here's how to do it.
+
+    The general idea is to nest elements to combine the animations: the container element will animate the zoom, and the image inside the container will be animated with a panning.
+    * Start by creating a div which will act as the image container. Set its style class to set `position:absolute`.
```
* s/to set/to/
* can you explain why we need to position the image absolute?

sebastianbenz commented on src/stories/20_Features/Visual_Effects.html:
```diff
@@ -0,0 +1,106 @@
+    The general idea is to nest elements to combine the animations: the container element will animate the zoom, and the image inside the container will be animated with a panning.
+    * Start by creating a div which will act as the image container. Set its style class to set `position:absolute`.
+    * Set its animation to be `animate-in="zoom-in"` or `animate-in="zoom-out"` and its duration with `"animate-in-duration=30s"`.
+    > In this demo we use 30 seconds but your mileage may vary. Try experimenting with different values to see what works best for your desired results.
```
You can create info boxes like this: 

```
  * Set its animation to be `animate-in="zoom-in"` or `animate-in="zoom-out"` and its duration with `"animate-in-duration=30s"`.
    <div class="ampstart-card info">
    **Tip:** In this demo we use 30 seconds but your mileage may vary. Try experimenting with different values to see what works best for your desired results.
    </div>
  ...
```

sebastianbenz commented on src/stories/20_Features/Animations.html:
```diff
@@ -212,6 +212,92 @@ <h1>whoosh-in-right</h1>
       </amp-story-grid-layer>
     </amp-story-page>
+    
+    <!-- ## Zoom-In -->
```
Perfect!  Thanks!

Enriqe commented:
Updated the demo + samples. They should work now, let me know!

sebastianbenz commented:
Works on mobile now :-), but no longer on desktop:

<img width="912" alt="screen shot 2018-03-22 at 19 26 21" src="https://user-images.githubusercontent.com/380472/37790517-17eacd20-2e07-11e8-83b2-a3c420282d8a.png">

sebastianbenz commented on src/stories/30_Visual_Effects/Ken_Burns.html:
```diff
@@ -0,0 +1,143 @@
+    -->
+    <amp-story-page id="ken-burns-effect2">
+        <amp-story-grid-layer template="vertical">
+          <div animate-in="zoom-in" animate-in-duration="30s" class="img-container">
```
I'd suggest making the animation a bit shorter (e.g. 5s). Having to wait 30s to see the effect is a bit long.

Enriqe replied:
The problem is that the nature of the effect is that it's a slow panning and zooming. If we use shorter times then the animation will run faster, beating the purpose of the effect. But I guess for the sake of showing the support of multiple images it makes sense.

sebastianbenz commented on src/stories/30_Visual_Effects/Ken_Burns.html:
```diff
@@ -0,0 +1,143 @@
+
+    <!-- ## Using Multiple Images -->
+    <!-- 
+      You can also achieve the effect with multiple images!
```
I had to see the demo to understand what was going on. Maybe change the intro to: 

"You can even implement a slideshow transitioning between multiple pages using this effect."

sebastianbenz commented on src/stories/30_Visual_Effects/Ken_Burns.html:
```diff
@@ -0,0 +1,143 @@
+              </amp-img>
+            </div>
+          </div>
+        </amp-story-grid-layer>
```
It'd be nice to show 1 or 2 more images - what do you think?

Enriqe replied:
Sounds good!

Enriqe commented:
> Works on mobile now :-), but no longer on desktop:

How are you testing this? It works on desktop for me locally.

sebastianbenz commented:
On desktop, you'll see a white frame once the vertical resolution is > 1024px.

Enriqe commented:
Oh I was just able to reproduce it. I'll update you when it's fixed.

sebastianbenz commented:
Hey Enriqe - any updates on this?

Enriqe commented:
Hey @sebastianbenz, sorry for the delay. I've just been able to do some adjustments and it seems to work for me in both desktop and mobile. Can you verify that it works on your side as well? Thanks!

Enriqe commented:
Hey Sebastian, were you able to verify it works on your side for desktop / mobile?

sebastianbenz commented:
Sorry, the problem still persists on desktop.

Enriqe commented:
This will be fixed with [a fix in the runtime side (14334)](https://github.com/ampproject/amphtml/pull/14334).

sebastianbenz commented on src/stories/20_Features/Animations.html:
```diff
@@ -258,57 +258,61 @@ <h1>zoom-out</h1>
-                 height="320">
-        </amp-img>
+      <amp-story-grid-layer template="fill">
+        <div>
```
Why do we need the div here?

Enriqe replied:
Since the `amp-story-grid-layer` template=fill only applies for the first element, we must wrap all elements inside the page inside an element. I added a small text regarding this on the ken burns page, I'll add it here as well.

sebastianbenz commented on src/stories/20_Features/Animations.html:
```diff
@@ -258,57 +258,61 @@ <h1>zoom-out</h1>
-                 width="720"
-                 height="320">
-        </amp-img>
+      <amp-story-grid-layer template="fill">
```
Can you re-add the title in the second layer?

Enriqe replied:
Sure 👍

sebastianbenz commented on src/stories/20_Features/Animations.html:
```diff
@@ -212,6 +222,111 @@ <h1>whoosh-in-right</h1>
+    <amp-story-page id="pan-left">
+      <amp-story-grid-layer template="fill">
+        <div>
+          <h1>pan-left</h1>
```
Stupid question: would it be more idiomatic to move the title into a different layer? e.g.:

```
<amp-story-grid-layer template="vertical">
   <h1>zoom-out</h1>
</amp-story-grid-layer>
```

Enriqe replied:
We could do it, but with a couple of caveats:

1. Ordering-wise in the markup, the layer where the title is located would have to be after the layer with the image, since the last "fill" template layer in an amp-story-page will be the element located at the top of the stacking context. 
2.  We would still need to wrap the amp-img in a div, even if it's the only element in the layer - since the "fill" template will change the dimensions of its first child, we need a div to "catch" those style changes. Only then will the image have its dimensions unchanged and the animations will work as intended.

I'll commit the changes to the PR so you can see how the code would look like.

sebastianbenz replied:
Ah - thanks for explaining!

In this sample I'd prefer a simpler markup using a vertical layout:

```
    <amp-story-page id="pan-left">
      <amp-story-grid-layer template="vertical">
        <h1>pan-left</h1>
        <amp-img animate-in="pan-left"
                 animate-in-duration="4s"
                 layout="fixed"
                 src="https://picsum.photos/720/320?image=1026"
                 width="720"
                 height="320">
        </amp-img>
      </amp-story-grid-layer>
    </amp-story-page>
```

The animations (in particular for pan-left and pan-right) are more visible as well (they move only a few pixels with the fill layout). The reason why I'm so picky about this is that this a sample demonstrating the animations, additional markup such as the div is distracting in this case. Wdyt?

Enriqe replied:
No problem!

Regarding the change to a vertical layout, if we take into account
[this new change on the runtime side](https://github.com/ampproject/amphtml/pull/14334), which will scale up the size of the image if it's smaller than the screen size, the image will sit on top of the title. Additionally, there will be some padding on the sides added by the vertical layout. (See screenshot). We could use a z-index to make the title be on top of the image, but not sure what you think about the padding. Let me know what you think.

![image](https://user-images.githubusercontent.com/5449100/38369561-02e16858-38b6-11e8-9680-8eb3335d7c5a.png)

![image](https://user-images.githubusercontent.com/5449100/38369617-2cc030e6-38b6-11e8-9602-4ad5b09b2e6e.png)

Enriqe commented:
Discussed with @newmuis offline and we will integrate a change on the runtime side [(same PR as above)](https://github.com/ampproject/amphtml/pull/14334) to make it possible to specify an `amp-img` as direct child of template fill and still work as intended. This with the goal that the animations are easy to use for publishers.

It's important to note that for multiple images publishers would still need to wrap them inside a div, and we consider that's okay for now because it's an "advanced" and very specific use case.

Enriqe commented:
The change is almost ready, so we can just wait until that is merged.

sebastianbenz commented:
Perfect!

On Mon, Apr 9, 2018 at 9:39 PM Enrique Marroquin <notifications@github.com>
wrote:

> The change is almost ready, so we can just wait until that is merged.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amp-by-example/pull/1202#issuecomment-379869740>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXOOBDcxhzQVvE2Tb2bXGVYuWI0Ai29ks5tm7jwgaJpZM4Syj-Q>
> .
>

Enriqe commented:
Change was submitted and I updated the sample.
